Fix post install & Pre-Alpha

n8marti opened the pull request:
Major changes:
- remove support for other GUIs besides Tk
- replace as many subprocesses as possible with pure python functions (including using non-standard-library modules)
- create binary using pyinstaller via `./scripts/build-binary.sh`
  - the installer binary also serves as the launcher binary
- create desktop launcher in postInstall
- read config file at start; update config file at end (only default location has been tested)

Minor changes:
- determine user Downloads folder from `~/.config/user-dirs.dirs`; fallback to `~/Downloads`
- add more logging output
- replace some terminal output functions with logging functions
- add some error handling for network timeouts (more to do still, though)

I don't know what your vision is for the post-install process, but I am proposing that the installer binary/script also be the command that launches the Logos app itself. So in this PR, if the binary/script is run (via the terminal or via the .desktop file), and if the LOGOS_EXE file exists and is executable, then it just attempts to start the app. If those conditions aren't met, then it transitions into the installation process. I'm open to other ideas, though--this is just what seemed logical to me.

n8marti committed:
fix a few syntax errors

n8marti committed:
fix subprocess errors due to 'command' built-in

n8marti committed:
change conflicting short options

n8marti committed:
use python3 explicitly

thw26 committed:
Initial Python conversion. Drop bash files.

thw26 committed:
Add TODO

PyInstaller

n8marti committed:
convert to logging module

Add Tk GUI; use config module for global variables (#7)

set default env at the start of main

add .gitignore

fix typo

add tk window

improve wineBinaryVersionCheck for wine < 9

fix config file label

send commands to check as a list

fix undefined variable by using env var

fallback to default config if Config file button cancelled

remove unused code

split some functions to utils module; auto-pick wine exe

split off more modules; successfully completes default install

disable Install button after click; convert to Exit at end

pass file names instead of handles to wine_reg_install

default DEBUG and VERBOSE to ''

give clearer output

add env/

use stdlib urllib instead of requests

Add TODO

PyInstaller

Add Tk GUI; use config module for global variables

fix reg file contents for winemenubuilder

use function to get wine-related env variables

use f-strings instead of format method

switch to using a config module instead of env variables

set vars in set_default_config

save time by not doing md5 hash

move vars to config module; fix typo

move wine-related functions to wine.py

remove unused createControlPanel

handle exception if install cancelled or fails

remove unused code

fix missing import; add dir path output to wait function

move set_appimage to utils.py

update imports after code changes

initial conversion to logging module

remove DEBUG and VERBOSE variables; remove DEBUG option

n8marti committed:
Add Tk GUI; use config module for global variables (#7)

set default env at the start of main

add .gitignore

fix typo

add tk window

improve wineBinaryVersionCheck for wine < 9

fix config file label

send commands to check as a list

fix undefined variable by using env var

fallback to default config if Config file button cancelled

remove unused code

split some functions to utils module; auto-pick wine exe

split off more modules; successfully completes default install

disable Install button after click; convert to Exit at end

pass file names instead of handles to wine_reg_install

default DEBUG and VERBOSE to ''

give clearer output

add env/

use stdlib urllib instead of requests

Add TODO

PyInstaller

Add Tk GUI; use config module for global variables

fix reg file contents for winemenubuilder

use function to get wine-related env variables

use f-strings instead of format method

switch to using a config module instead of env variables

set vars in set_default_config

save time by not doing md5 hash

move vars to config module; fix typo

move wine-related functions to wine.py

remove unused createControlPanel

handle exception if install cancelled or fails

remove unused code

fix missing import; add dir path output to wait function

move set_appimage to utils.py

update imports after code changes

n8marti committed:
add an explicit timeout to https requests

add an explicit timeout to https requests

cleanup logging

remove GUI variable check b/c logic is handled elsewhere

modify logos_progress to use dots for now

add a newline at the end of config file

remove check for VERBOSE variable

ignore build dir

n8marti committed:
config conversion cleanup

add FIXME question re: wineboot

fix variable assignment from config file

ensure config file is properly written and properly read

fix variable name error

fix variable names

n8marti committed:
block wine-mono window during install

n8marti committed:
conversion to pure python cleanup

use WINEPREFIX var directly

convert subprocesses to pure python

add short delay during wait loop

move die and parse functions up before getDialog

n8marti committed:
use XDG_CONFIG_HOME to get Downloads folder

n8marti committed:
strip quotes from path string

n8marti committed:
conversion to psutil

use psutil to find and wait for PID using directory

replace several subprocess calls with psutil

n8marti committed:
rework to prep for pyinstaller build

large-scale rework to prep for pyinstaller build

update module imports

n8marti committed:
Merge branch 'python' into py-add-logging

thw26 committed:
Merge pull request #8 from n8marti/py-add-logging

Convert to using logging module

n8marti committed:
ignore build dir

n8marti committed:
Merge branch 'python' of https://github.com/thw26/LogosLinuxInstaller into python

n8marti committed:
add option to delete installer log

n8marti committed:
logging cleanup

remove references to VERBOSE var

fix wrong var name

cleanup wait_on and logos_progress

cleanup terminal and log file output

cleanup comments

cleanup winetricks_install

remove duplicated function calls

more terminal and log output cleanup

fix logging function error

cleanup run_wine_proc function

add debug output

cleanup logos_setup function

add dep samba-winbind-clients

add distro names for naive devs

cleanup & improve error handling

standardize exit to sys.exit

standardize exit as sys.exit

standardize exit to sys.exit

remove logos_info function

update pyinstaller check

handle network timeout while downloading releases

n8marti committed:
improve logging

improve logging

add more logging output

check ~/Downloads for Logos icon file

improved logging

remove unused imports

specify MSI file name that is executed

n8marti committed:
updated build

n8marti committed:
replace wget with requests-based net_get function

remove q parameter from same_size

fix error if file not given

use requests to get UrlProps

add and use net_get function based on requests module

add net_get function that uses requests

use net_get in tk app

add write_progress_bar function

allow target to be None

get Logos releases using net_get

use net_get for cli_download

verify file size in logos_reuse_download

use cli_download; remove logos_download

remove unused imports

add md5 check for amazons3 files

handle ctrl+c interrupt

add md5 check to downloaded file check

use logos_reuse_download for Logos installer

cleanup file verification functions

do both size and md5 checks on downloaded files

thw26 commented on dependencies.md:
Does this file get used by the installer, or is this for reference?

n8marti replied:
Oh, this was only for my reference as I was trying to keep track of details. We probably don't need to keep it, really.

thw26 replied:
Okay. If anything, push those updates to the README or to the OS-specific dependencies in the installer :)

thw26 commented on installer.py:
```diff
@@ -115,41 +98,34 @@ def chooseProduct():
         config.LOGOS_RELEASE_VERSION = logos_release_version
     else:
-        logging.critical("Failed to fetch LOGOS_RELEASE_VERSION.")
+        logos_error("Failed to fetch LOGOS_RELEASE_VERSION.")
```
Is this right? Weren't we changing to logging.x?

thw26 replied:
This of course could just be me misremembering the logging changes.

n8marti replied:
logos_error now uses logging to record the main error message in the log file and display it on stderr. But I kept the function because it also gives the "get help" text and exits with an error code. I thought it seemed fitting for critical errors.

thw26 replied:
Sounds good!

thw26 reviewed (commented):
Most of this looks good. I think my only outstanding question, besides the comment (which could be me not remembering all the changes with the Great File Split, is the one shared in chat, and unfortunately I have not yet had time to run an actual install test using all your changes.

If I am understanding correctly, will GUI users still be able to run the menu commands after an install, or does the change to launch an EXE if found negate that ability? This could just be misunderstanding on my part.

If so, I think I favor "creating" two shortcuts, one to our launcher program, which is just keeping the binary around, and one to Logos itself, which we already do.

n8marti committed:
properly decode wine cmd output

n8marti committed:
add ControlPanel window

move control panel function to main module

add and update functions for toggling app logging

move control panel functions to control.py

add ControlPanel window and update related functions

create 2 desktop launchers

n8marti committed:
update binary build

thw26 commented on installer.py:
```diff
@@ -407,135 +393,192 @@ def installLogos10():
+        logos_error(f"Installation failed. {config.LOGOS_EXE} not found. Exiting…\nThe {config.FLPRODUCT} executable was not found. This means something went wrong while installing {config.FLPRODUCT}. Please contact the Logos on Linux community for help.")
 
 def install():
-    # BEGIN PREPARATION
+    prepare_install()
+    finish_install()
+
+def prepare_install():
     chooseProduct()  # We ask user for his Faithlife product's name and set variables.
     chooseVersion()  # We ask user for his Faithlife product's version, set variables, and create project skeleton.
     getLogosReleaseVersion(getLogosReleases())
     logos_setup() # We set some basic variables for the install, including retrieving the product's latest release.
     chooseInstallMethod()  # We ask user for his desired install method.
```
I think these two functions need to be flipped, right?

n8marti replied:
Maybe the names are misleading and were ill-chosen, but the order is correct. Basically, `prepare_install()` sets all the installer choices, then `finish_install()` applies everything. So curses calls `install()`, which then calls `prepare_install()` and `finish_install()`, but the Tk window sets all the choices in the GUI window then calls `finish_install()` directly. It doesn't use `install()` or `prepare_install()` at all.

thw26 replied:
Ah, in this case, I meant are they defined in proper order. I would have assumed prepare_install would need to be defined before install.

n8marti commented:
> If I am understanding correctly, will GUI users still be able to run the menu commands after an install, or does the change to launch an EXE if found negate that ability? This could just be misunderstanding on my part.

I'm not sure if the addition of the Control Panel app answers this question for you or not. As it stands, a GUI-only user has two choices after a successful install: 1) Run the installed app directly, which happens via the "LogosBible.desktop" file; or 2) Run the Control Panel app via the "LogosBible-ControlPanel.desktop" file.

There's not currently a "Delete installation" or "Reinstall" option in the Control Panel app, but if either of those options were provided, then the GUI-only user _could_ go back through the original installation options by essentially starting over. But if they wanted to install a 2nd wine prefix for a test installation, that would currently have to be done via the CLI app.

thw26 commented:
> > If I am understanding correctly, will GUI users still be able to run the menu commands after an install, or does the change to launch an EXE if found negate that ability? This could just be misunderstanding on my part.
> 
> I'm not sure if the addition of the Control Panel app answers this question for you or not. As it stands, a GUI-only user has two choices after a successful install: 1) Run the installed app directly, which happens via the "LogosBible.desktop" file; or 2) Run the Control Panel app via the "LogosBible-ControlPanel.desktop" file.
> 
> There's not currently a "Delete installation" or "Reinstall" option in the Control Panel app, but if either of those options were provided, then the GUI-only user _could_ go back through the original installation options by essentially starting over. But if they wanted to install a 2nd wine prefix for a test installation, that would currently have to be done via the CLI app.

The control panel app resolves my concern.

n8marti committed:
update requirements with requests & its deps

n8marti committed:
pre-alpha cleanup

pre-alpha cleanup

don't include args in app launch after install

use logos_reuse_download to get icon

remove unused imports

update binary

configure logging sooner in script

remove unneeded file

n8marti committed:
fix to get correct WINEBIN_CODE in GUI
